### ISSUE TITLE
handle dartdoc references to non-imported symbols (was: dartdoc-specific imports)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/dartdoc#1153]
Sometimes, a library, like the Flutter rendering library, wants to refer to another library, like the Flutter widgets library, in the documentation, but doesn't want to introduce a dependency in the code. In this case, for example, we definitely do not want any sort of runtime relationship from the rendering library to the widgets library.

We can't use `import` for this because it risks code accidentally depending through the import.

One possible solution would be to have something like `/// @import` which would cause the dartdoc code to import it. Another would be for the dartdoc tool to be able to receive a list of imports to imply at the top of every file. I don't really mind how we solve this.

**Comment by pq** [comment]
cc @keertip for thoughts.

**Comment by pq** [comment]
Here's an example as food for thought from [rendering.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/rendering.dart):

```
/// The Flutter rendering tree.
/// 
/// To use, import `package:flutter/rendering.dart`.
///
/// The [RenderObject] hierarchy is used by the Flutter Widgets
/// library to implement its layout and painting back-end. Generally,
/// while you may use custom [RenderBox] classes for specific effects
/// in your applications, most of the time your only interaction with
/// the [RenderObject] hierarchy will be in debugging layout issues.
///
/// If you are developing your own library or application directly on
/// top of the rendering library, then you will want to have a binding
/// (see [BindingBase]). You can use [RenderingFlutterBinding], or you
/// can create your own binding. If you create your own binding, it
/// needs to import at least [SchedulerBinding], [GestureBinding],
/// [ServicesBinding], and [RendererBinding]. The rendering library
/// does not automatically create a binding, but relies on one being
/// initialized with those features.
```

lints to:

```
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 13, col 10)
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 17, col 11)
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 17, col 39)
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 19, col 31)
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 19, col 51)
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 20, col 6)
[lint] Only reference in scope identifiers in doc comments. (packages/flutter/lib/rendering.dart, line 20, col 29)
```

**Comment by devoncarew** [comment]
The latest thinking on this is to support something like fully qualified symbol references in square brackets (dartdoc specific imports might get a little hairy). Something like `[widgets.Draggable]` or `[package:flutter/widgets Draggable]`.

/cc @pq @brianwilkerson

**Comment by pq** [comment]
👍 I'm in favor in taking the cue from Java's qualified names.

Curious where you weigh in @bwilkerson ?

**Comment by Hixie** [comment]
I could totally work with fully-qualified identifiers in square brackets (assuming they don't render that way, at least).

**Comment by pq** [comment]
@bwilkerson, @scheglov : how does this proposal sit with you?

**Comment by scheglov** [comment]
I like the idea of fully qualified names.

**Comment by eseidelGoogle** [comment]
FWIW this came up again tonight.  https://docs.flutter.io/flutter/material/Icons-class.html  Icon and IconButton (the classes which consume the data in icons.dart) can't be linked since they're not imported.

**Comment by pq** [comment]
@scheglov : am I right that this wouldn't be too hard to add?

**Comment by scheglov** [comment]
Yes, I think it wouldn't be too hard.

**Comment by eseidelGoogle** [comment]
In case it's helpful, here is the output of running flutter analyze with the comment_references lint on.  Currently its disabled with a comment saying it's blocked by this bug:
https://gist.github.com/eseidelGoogle/618570930fb1752e0f5bc7afa29aae91

I don't think this is a huge deal by any means, just wanted you to have full context.

**Comment by qchong** [comment]
We're also getting feedback from the UX team that the missing links in the Flutter API docs is a usability problem.

@scheglov Ping. Is there a plan to fix this issue?

**Comment by scheglov** [comment]
I don't have any specific plans about fixing this issue yet.

**Comment by sethladd** [comment]
If we fix this, we'll probably also help crossdart.info link to more names. Right now, I can't link to Widget from Button's crossdart page: https://www.crossdart.info/p/flutter/0.0.21/src/material/button.dart.html

I'd imagine that Widget would be a link.

**Comment by Hixie** [comment]
That last thing seems like a separate bug. The code knows what Widget it so the hyperlink should work. This bug is just about hyperlinks from dartdocs themselves to identifiers that are otherwise not in scope.

**Comment by sethladd** [comment]
I think @devoncarew and @keertip have an idea of how to handle this? Comments here can help speed up implementation :)

**Comment by devoncarew** [comment]
I believe where this was left:

- for resolving references to symbols that are not currently imported (think referencing a material component from the flutter widget layer), we would need those dartdoc referenced to be fully qualified - to include enough information for us to resolve them. Something like `[package:flutter/material.dart MaterialApp]`
- when the analyzer is not able to resolve a symbol reference in some dartdoc, dartdoc itself will try to resolve; it should be able to locate anything currently being documented in that session (so, anything in the current context). It the above example, it would resolve to `material/MaterialApp-class.html`

**Comment by kevmoo** [comment]
That would rock!

I'd argue that we'd want to avoid too much (or any?) logic in dartdoc that
does magic lookup. We'd want the editing experience to clearly align with
what dartdoc does. We want the link to warn if they are referencing
something that doesn't exist, etc...

**Comment by sethladd** [comment]
> We want the link to warn if they are referencing something that doesn't exist

Is there an issue open on the analyzer for this "extended link" syntax and support?

**Comment by devoncarew** [comment]
There isn't an issue - putting the logic into dartdoc was the resolution in order to avoid implementation complexity in the analyzer and analysis server (see https://github.com/dart-lang/sdk/issues/27471#issuecomment-251179606).

**Comment by Hixie** [comment]
I really would like to avoid having to fully qualify every reference. That would make the dartdocs unreadable, and it turns out that people look at the source a LOT. (Yay jump to definition.)

If there is only one identifier with the given name in the entire session, why not just assume that's what it means? And if there are duplicates, then require disambiguation. If we do this then we can even have it fail the entire process if any reference can't be resolved.

**Comment by devoncarew** [comment]
So, `[MaterialApp]`, which would work if MaterialApp was unique, and `[package:flutter/material.dart MaterialApp]` (or similar fully-qualified syntax) if there was a need to disambiguate?

**Comment by Hixie** [comment]
That would be sufficiently concise for me I think.

**Comment by astashov** [comment]
I'm looking into it

**Comment by astashov** [comment]
Created initial PR for that. There's still some outstanding questions:

* What to do in case of ambiguity here? Show a warning, crash with an error message, ignore? Doesn't seem like super critical thing, so I guess a warning would work, but would love to hear other arguments :)
* Do we want to show short name in case a fully qualified name is provided? I saw in the comments in the ticket (https://github.com/dart-lang/dartdoc/issues/1153#issuecomment-236012803) that @Hixie wants that, right?

Thanks!

**Comment by Hixie** [comment]
In case of ambiguity, I would fail hard.
Ideally we'd have analyzer warnings that let us know this is going to happen.

Please always give just the short name in the output, yes.

BTW once we have this we'll ideally also want to fail hard if there's every a [link] that is not resolvable. (Again, ideally, with an analyzer warning to let us know it'll happen before we try to run the docs.)

**Comment by sethladd** [comment]
Thanks!!

**Comment by gnprice** [comment]
For cross-reference: the idea early in this thread of having a way to make imports only for dartdoc purposes is now tracked at:
* https://github.com/dart-lang/sdk/issues/44637

and has a proposed design at:
* https://github.com/dart-lang/sdk/issues/50702
